skrub/_parser.py

```diff
@@ -0,0 +1,387 @@
+    """Parse datetime columns of a Pandas or Polars dataframe.
+
+    Parameters
+    ----------
+    See the docstring of skrub._datetime_encoder.to_datetime.
+
+    Attributes
+    ----------
+    inferred_column_formats_ : mapping of string to string or DateTime64DType.
+        Mapping between column names infered  and their format.
+        If the column is already of datetime dtype, we use this as a format.
+    """
+
+    def __init__(
+        self, format=None, errors="coerce", random_state=None, **to_datetime_params
+    ):
+        super().__init__(errors=errors)
+        self.format = format
+        self.random_state = random_state
+        self.to_datetime_params = to_datetime_params
+
+    def fit(self, X, y=None):
+        del y
+        self.inferred_column_formats_ = dict()
+        return super().fit(X)
+
+    def _infer(self, column_name, column):
+        column = column.to_numpy()  # TODO: remove
+        column = column[pd.notnull(column)]
+
+        if is_numeric_dtype(column):
+            return None
+
+        elif is_datetime64_any_dtype(column):
+            self.inferred_column_formats_[column_name] = None
+            return column.dtype
+
+        elif _is_column_datetime_parsable(column):
+            # _guess_datetime_format only accept string columns.
+            # We need to filter out columns of object dtype that
+            # contains e.g., datetime.datetime or pd.Timestamp.
+            column_str = column.astype(str)
+            if np.array_equal(column, column_str):
+                datetime_format = _guess_datetime_format(column, self.random_state)
+            else:
+                # We don't need to specify a parsing format
+                # for columns that are already of type datetime64.
+                datetime_format = None
+            self.inferred_column_formats_[column_name] = datetime_format
+            return column.dtype
+
+        return None
+
+    def _parse(self, column_name, column):
+        # In Pandas 2.1, pd.to_datetime has a different behavior
+        # when the items of column are np.str_ instead of str.
+        # TODO: find the bug and remove this line
+        column = list(map(str, column))
+        datetime_format = self.format or self.inferred_column_formats_[column_name]
+        return pd.to_datetime(
+            column,
+            format=datetime_format,
+            errors=self.errors,
+            **self.to_datetime_params,
+        )
+
+
+class _NumericParser(_BaseParser):
+    """Parse numeric columns of a Pandas dataframe."""
+
+    def _infer(self, column_name, column):
+        if is_datetime64_any_dtype(column):
+            return None
+
+        try:
+            return pd.to_numeric(column, errors="raise").dtype
+        except (ValueError, TypeError):
+            return None
+
+    def _parse(self, column_name, column):
+        return pd.to_numeric(column, errors=self.errors)
+
+
+def _union_category(column, dtype):
```

Comment:
I think it might be best if transform didn't modify the state of the transformer and if the output type of transform was always the same (rather than different categorical dtypes). maybe an "unknown" category could be systematically added and unseen categories would go in there?